Automatically update map when hosting a game

Marc-Spector opened the pull request:
Closes #1986

IvanPavilionG7 committed:
update an outdated map when hosting a game

IvanPavilionG7 committed:
fix codacy issues

Marc-Spector commented:
Do I need to add unit tests? What do you think?

Sheikah45 commented:
Tests are always good. Helps to make sure that things don't get broken by future changes

IvanPavilionG7 committed:
add some unit tests

IvanPavilionG7 committed:
fix null point exception in a test method

IvanPavilionG7 committed:
add ignore annotation to an empty test method

Sheikah45 commented on src/main/java/com/faforever/client/map/MapBean.java:
```diff
@@ -243,6 +243,20 @@ public int compareTo(@NotNull MapBean o) {
     return getDisplayName().compareTo(o.getDisplayName());
   }
 
+  public int compareByVersion(@NotNull MapBean o) throws CompareMapVersionException {
+    if (!this.getDisplayName().equalsIgnoreCase(o.getDisplayName())) {
+      throw new CompareMapVersionException("cannot compare map versions with different map names");
+    }
+
+    ComparableVersion v1 = this.getVersion();
+    ComparableVersion v2 = o.getVersion();
+    if (v1 == null || v2 == null) {
+      throw new CompareMapVersionException(String.format("cannot compare map versions, map '%s' has null version",
+          v1 == null ? getFolderName() : o.getFolderName()));
+    }
+    return v1.compareTo(v2);
+  }
+
```
I would move this operation to the MapService as the MapBean is really just a data storage object

Sheikah45 replied:
Also the latestMapVersion should always be the most recent so you don't have to check the version specifically just check if the latest is the same as the one you are comparing to.

Marc-Spector replied:
Good! I did not think about it.

IvanPavilionG7 committed:
move the compare version method to MapService class and rename it
to "isNewVersionMap"
remove MapBeanTest class

IvanPavilionG7 committed:
add unit test
fix codacy issues

IvanPavilionG7 committed:
add new condition for official maps when hosting a game

IvanPavilionG7 committed:
add new unit tests

IvanPavilionG7 committed:
add unit tests

IvanPavilionG7 committed:
remove unused import

Sheikah45 commented on src/main/java/com/faforever/client/map/MapService.java:
```diff
@@ -475,6 +482,60 @@ public void evictCache() {
+          if (isNewVersionMap(latestMap, map)) {
+            return Optional.of(latestMap);
+          }
+        } catch (CompareMapVersionException ignored) {  }
```
should at least add a log statement

Sheikah45 reviewed (changes requested):
Very thorough tests!
Looks good just a log statement in the ignored exception would be nice

IvanPavilionG7 committed:
add log statement when compare map versions exception
add new unit tests
improve code for creating game (for any error when checking or
updating map, create a game always)

IvanPavilionG7 committed:
add body to empty unit test method
